src/Core/Termination/CallGraph.idr

```diff
@@ -145,49 +146,56 @@ mutual
                     else pure $ Ref fc Func n
         conIfGuarded tm = pure tm
 
-  -- Expand the size change argument list with 'Nothing' to match the given
-  -- arity (i.e. the arity of the function we're calling) to ensure that
-  -- it's noted that we don't know the size change relationship with the
-  -- extra arguments.
-  expandToArity : Nat -> List (Maybe (Nat, SizeChange)) ->
-                  List (Maybe (Nat, SizeChange))
-  expandToArity Z xs = xs
-  expandToArity (S k) (x :: xs) = x :: expandToArity k xs
-  expandToArity (S k) [] = Nothing :: expandToArity k []
+  knownOr : SizeChange -> Lazy SizeChange -> SizeChange
+  knownOr Unknown y = y
+  knownOr x _ = x
+
+  plusLazy : SizeChange -> Lazy SizeChange -> SizeChange
+  plusLazy Smaller _ = Smaller
+  plusLazy x y = x |+| y
 
   -- Return whether first argument is structurally smaller than the second.
-  smaller : Bool -> -- Have we gone under a constructor yet?
-            Defs ->
-            Maybe (Term vars) -> -- Asserted bigger thing
-            Term vars -> -- Term we're checking
-            Term vars -> -- Argument it might be smaller than
-            Bool
-  smaller inc defs big _ (Erased _ _) = False -- Never smaller than an erased thing!
+  sizeCompare : Term vars -> -- RHS: term we're checking
+                Term vars -> -- LHS: argument it might be smaller than
+                SizeChange
```

Comment:
I very much agree with this but don't have any good ideas how to do it. It might be possible to classify the syntactic fragment that these "patterns" inhabit but I don't think it would be straight-forward.